File: src/mode.c
#define R_NO_REMAP

void mode_c(
    double *input,    // Input vector
    int *n,           // Length of 'input'
    double *na_flag,  // 'NA' flag 
    double *result    // Result (length 1)
    ) {

    int i, j;
    int max_count = 0;

    // Find mode
    for(i = 0; i < *n; i++) {
        int count = 0;
        for(j = 0; j < *n; ++j) {
            if(input[j] != *na_flag && input[j] == input[i]) count++;
        }
        if(count > max_count) {
            max_count = count;
            result[0] = input[i];
        }
    }

    // Find all values sharing same 'max_count'
    // ...

    // Select final result
    // ...

}

File: R/mode.R
#' Mode
#'
#' Find the mode (i.e., most common value) in a numeric vector. In case of ties, the first encountered value is returned.
#' 
#' @param x A \code{numeric} or \code{logical} vector
#' @param na_flag Value used to mark \code{NA} values in C code. This should be set to a value which is guaranteed to be absent from the input vector \code{x} (default is \code{-9999}).
#' 
#' @return The mode, \code{numeric} vector of length 1
#'
#' @examples
#' x = c(3, 2, 5, 5, 3, 10, 2, 5)
#' mode_value(x)
#' 
#' @export

mode_value = function(x, na_flag = -9999) {

  # Checks
  stopifnot(is.numeric(x) | is.logical(x))
  stopifnot(is.numeric(na_flag))
  stopifnot(length(na_flag) == 1)

  # Replace 'NA' with 'na_flag'
  x[is.na(x)] = na_flag

  # Find mode
  result = .C(
    "mode_c",
    as.double(x),
    as.integer(length(x)),
    as.double(na_flag),
    as.double(na_flag)
  )
  result = result[[length(result)]]

  # Replace 'na_flag' with 'NA'
  if(result == na_flag) result = NA

  # Return
  return(result)

}

